### PR TITLE
remote: enhance FetchByDigest() to handle more mediatypes

### PR DESCRIPTION
**Pull request by jiangliu**
Enhance FetchByDigest() to handle manifest with mediatype "application/vnd.oci.image.manifest.v1+json"

Fixes: https://github.com/containerd/nydus-snapshotter/issues/474

### CODE REVIEW

**Comment by liubogithub** [comment]
Looks good to me.

File: pkg/remote/remotes/docker/fetcher.go [hunk 1]
```diff
@@ -209,7 +215,22 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest) (i
...
+				firstErr = err
+			}
+		}
+	}
+
+	if getReq == nil {
+		// Fall back to the "manifests" endpoint with specific mediatype
+		mediatype = "application/vnd.oci.image.manifest.v1+json"
+		for _, host := range r.hosts {
+			getReq, sz, err = r.createGetReq(ctx, host, mediatype, "manifests", dgst.String())
```
Comment (imeoer) [inline pkg/remote/remotes/docker/fetcher.go:233]: Can we make a slice like `mediaTypes := []string{ images.MediaTypeDockerSchema2Manifest, images.MediaTypeDockerSchema2ManifestList, images.MediaTypeDockerSchema1Manifest, ocispec.MediaTypeImageManifest, ocispec.MediaTypeImageIndex }` ? And then add them into the `Accept` request header.

File: pkg/remote/remotes/docker/fetcher.go [hunk 2]
```diff
@@ -151,11 +151,16 @@ func (r dockerFetcher) Fetch(ctx context.Context, desc ocispec.Descriptor) (io.R
 	})
 }
 
-func (r dockerFetcher) createGetReq(ctx context.Context, host RegistryHost, ps ...string) (*request, int64, error) {
+func (r dockerFetcher) createGetReq(ctx context.Context, host RegistryHost, mediatype string, ps ...string) (*request, int64, error) {
```
Comment (changweige) [inline pkg/remote/remotes/docker/fetcher.go:154]: The whole remote package is copied from containerd. Why containerd does not have this problem?

File: pkg/remote/remotes/docker/fetcher.go (hunk 1 above)
Comment (jiangliu) [reply pkg/remote/remotes/docker/fetcher.go:224]: Fixed

File: pkg/remote/remotes/docker/fetcher.go (hunk 2 above)
Comment (jiangliu) [reply pkg/remote/remotes/docker/fetcher.go:154]: It does have this issue, only no client to trigger it:)

**Comment by cardyok** [comment]
Maybe we should simply change the argument for FetchByDigest, use descriptor instead of digest and use the mediatype defined by client? Since original purpose of FetchByDigest is to use registry as a general purpose CAS, it makes more sense to avoid defining image related headers.

File: pkg/remote/remotes/docker/fetcher.go [hunk 3]
```diff
@@ -208,8 +214,14 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest) (i
 
 	if getReq == nil {
 		// Fall back to the "manifests" endpoint
+		mediatype = strings.Join([]string{
```
Comment (imeoer) [inline pkg/remote/remotes/docker/fetcher.go:217]: Better rename to `mediaTypes`.

File: pkg/remote/remotes/docker/fetcher.go [hunk 4]
```diff
@@ -208,8 +214,14 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest) (i
...
+			images.MediaTypeDockerSchema2Manifest,
+			images.MediaTypeDockerSchema2ManifestList,
+			ocispec.MediaTypeImageManifest,
+			ocispec.MediaTypeImageIndex,
+		}, ",")
```
Comment (imeoer) [inline pkg/remote/remotes/docker/fetcher.go:222]: `", "`

File: pkg/remote/remotes/docker/fetcher.go (hunk 3 above)
Comment (jiangliu) [reply pkg/remote/remotes/docker/fetcher.go]: fixed

File: pkg/remote/remotes/docker/fetcher.go (hunk 4 above)
Comment (jiangliu) [reply pkg/remote/remotes/docker/fetcher.go]: fixed

**Comment by jiangliu** [comment]
> Maybe we should simply change the argument for FetchByDigest, use descriptor instead of digest and use the mediatype defined by client? Since original purpose of FetchByDigest is to use registry as a general purpose CAS, it makes more sense to avoid defining image related headers.

We only have the manifest digest and not known the mediaType, so need to probe the mediaType.

File: pkg/remote/remotes/docker/fetcher.go (hunk 2 above)
Comment (imeoer) [inline pkg/remote/remotes/docker/fetcher.go:154]: `mediaTypes string`

File: pkg/remote/remotes/docker/fetcher.go [hunk 5]
```diff
@@ -190,13 +195,14 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest) (i
...
-		firstErr error
+		getReq    *request
+		sz        int64
+		firstErr  error
+		mediaType string
 	)
 
 	for _, host := range r.hosts {
-		getReq, sz, err = r.createGetReq(ctx, host, "blobs", dgst.String())
+		getReq, sz, err = r.createGetReq(ctx, host, mediaType, "blobs", dgst.String())
```
Comment (imeoer) [inline pkg/remote/remotes/docker/fetcher.go:205]: `mediaTypes`

**Review by imeoer — COMMENTED** [review]
others LGTM, in addition, it is better to comment that we have modified this codes, and wait for upstream to accept it.

File: pkg/remote/remotes/docker/fetcher.go (hunk 2 above)
Comment (jiangliu) [reply pkg/remote/remotes/docker/fetcher.go:154]: let's keep consistence with open(), which uses `mediatype`?

**Review by imeoer — APPROVED** [review]

File: pkg/remote/remotes/docker/fetcher.go (hunk 2 above)
Comment (imeoer) [reply pkg/remote/remotes/docker/fetcher.go:154]: okay

**Review by imeoer — APPROVED** [review]

**Pull request merged**